### PR TITLE
[#2209] feat(server): Support config buffer size for LocalFileWriter

### PR DESCRIPTION
**Pull request by lwllvyb**
### What changes were proposed in this pull request?

Support config buffer size for LocalFileWriter

### Why are the changes needed?

Fix: #2209 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Locally test.

### CODE REVIEW

File: storage/src/main/java/org/apache/uniffle/storage/handler/impl/LocalFileWriter.java [hunk 1]
```diff
@@ -33,10 +35,16 @@ public class LocalFileWriter implements FileWriter, Closeable {
   private FileOutputStream fileOutputStream;
   private long nextOffset;
 
+  @VisibleForTesting
   public LocalFileWriter(File file) throws IOException {
+    this(file, 8 * 1024);
+  }
+
+  public LocalFileWriter(File file, long bufferSize) throws IOException {
```
Comment (maobaolong) [inline storage/src/main/java/org/apache/uniffle/storage/handler/impl/LocalFileWriter.java:43]: Why you define long type but used as int type? I guess it is from the long type configure item, so you can convert to int type earlier, it is better to convert it after get value from config.

**Review by maobaolong — COMMENTED** [review]
this is a performance improvement changes, It LGTM overall, left a minor comment inline. BTW, it would be nice if you paste the performance test result as the description or a link of this PR.

File: storage/src/main/java/org/apache/uniffle/storage/handler/impl/LocalFileWriter.java (hunk 1 above)
Comment (lwllvyb) [reply storage/src/main/java/org/apache/uniffle/storage/handler/impl/LocalFileWriter.java:43]: BufferedOutputStream's constructor need int type and getSizeAsBytes will convert the string value to long type.
I will change the code to convert the value's type earlier.

**Review by maobaolong — APPROVED** [review]
@lwllvyb Now, this looks clear, LGTM

File: common/src/main/java/org/apache/uniffle/common/config/RssBaseConf.java [hunk 2]
```diff
@@ -294,6 +294,18 @@ public class RssBaseConf extends RssConf {
                   + " first combining the username and the password with a colon (uniffle:uniffle123)"
                   + ", and then by encoding the resulting string in base64 (dW5pZmZsZTp1bmlmZmxlMTIz).");
 
+  public static final ConfigOption<String> RSS_STORAGE_WRITE_DATA_BUFFER_SIZE =
+      ConfigOptions.key("rss.storage.write.data.buffer.size")
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/config/RssBaseConf.java:298]: The last segment prefer using Camel style like `rss.storage.write.dataBufferSize`.

File: common/src/main/java/org/apache/uniffle/common/config/RssBaseConf.java [hunk 3]
```diff
@@ -294,6 +294,18 @@ public class RssBaseConf extends RssConf {
...
+          .stringType()
+          .defaultValue("8k")
+          .withDescription("The buffer size to cache the write data content.");
+
+  public static final ConfigOption<String> RSS_STORAGE_WRITE_INDEX_BUFFER_SIZE =
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/config/RssBaseConf.java:303]: ditto.

**Comment by lwllvyb** [comment]
Per event { num: 1000, blockSize: 16384 } eventNum: 1000 total size: 16384000000
write: 			75860 72293 73507 73585 73769 Average: 73802 ms, opt: 1.0
writeBuffer(1MB): 	68339 69674 67826 68017 69397 Average: 68650 ms, opt: 0.9301916
writeBuffer(8MB): 	73166 71645 73885 70285 72046 Average: 72205 ms, opt: 0.978361
writeBuffer(16MB): 	71575 70121 70922 70513 70628 Average: 70751 ms, opt: 0.95865965

Per event { num: 1000, blockSize: 32768 } eventNum: 500 total size: 16384000000
write: 			35487 35971 36248 37756 38809 Average: 36854 ms, opt: 1.0
writeBuffer(1MB): 	33903 33720 35004 33963 33985 Average: 34115 ms, opt: 0.9256797
writeBuffer(8MB): 	36286 36067 36605 34224 34378 Average: 35512 ms, opt: 0.96358603
writeBuffer(16MB): 	33795 33789 34116 35614 34370 Average: 34336 ms, opt: 0.9316763

Per event { num: 1000, blockSize: 65536 } eventNum: 250 total size: 16384000000
write: 			18000 18853 18096 17837 18037 Average: 18164 ms, opt: 1.0
writeBuffer(1MB): 	16610 16549 16420 16409 16759 Average: 16549 ms, opt: 0.9110879
writeBuffer(8MB): 	17397 17353 17316 17487 16998 Average: 17310 ms, opt: 0.9529839
writeBuffer(16MB): 	17518 16970 17526 16971 17320 Average: 17261 ms, opt: 0.95028627

Per event { num: 1000, blockSize: 262144 } eventNum: 125 total size: 32768000000
write: 			9059 9083 9162 8962 9588 Average: 9170 ms, opt: 1.0
writeBuffer(1MB): 	8283 8254 8221 8187 8305 Average: 8250 ms, opt: 0.89967287
writeBuffer(8MB): 	8475 8376 8358 8303 8363 Average: 8375 ms, opt: 0.91330427
writeBuffer(16MB): 	8382 8476 8390 8489 8470 Average: 8441 ms, opt: 0.92050165

Per event { num: 1000, blockSize: 524288 } eventNum: 62 total size: 32505856000
write: 			4394 4437 4429 4534 4399 Average: 4438 ms, opt: 1.0
writeBuffer(1MB): 	4023 4042 4162 4199 4199 Average: 4125 ms, opt: 0.92947274
writeBuffer(8MB): 	4178 4260 4081 4109 4294 Average: 4184 ms, opt: 0.942767
writeBuffer(16MB): 	4114 6990 4580 4227 4186 Average: 4819 ms, opt: 1.0858495

Per event { num: 1000, blockSize: 1048576 } eventNum: 31 total size: 32505856000
write: 			2237 2235 2250 2346 2386 Average: 2290 ms, opt: 1.0
writeBuffer(1MB): 	2082 2007 2034 2098 2019 Average: 2048 ms, opt: 0.89432317
writeBuffer(8MB): 	2056 2204 2076 2085 2095 Average: 2103 ms, opt: 0.9183406
writeBuffer(16MB): 	2105 2163 2078 2088 2145 Average: 2115 ms, opt: 0.92358077

Per event { num: 1000, blockSize: 2097152 } eventNum: 15 total size: 31457280000
write: 			1180 1163 1240 1216 1158 Average: 1191 ms, opt: 1.0
writeBuffer(1MB): 	989 958 980 973 997 Average: 979 ms, opt: 0.8219983
writeBuffer(8MB): 	992 1004 1013 997 979 Average: 997 ms, opt: 0.83711165
writeBuffer(16MB): 	999 988 1017 1022 1026 Average: 1010 ms, opt: 0.8480269

After some benchmarks with different block sizes and different buffer sizes, the performance results are 5% ~ 18% improvement.

File: common/src/main/java/org/apache/uniffle/common/config/RssBaseConf.java (hunk 2 above)
Comment (lwllvyb) [reply common/src/main/java/org/apache/uniffle/common/config/RssBaseConf.java]: I have changed the code as you suggested.

File: common/src/main/java/org/apache/uniffle/common/config/RssBaseConf.java (hunk 3 above)
Comment (lwllvyb) [reply common/src/main/java/org/apache/uniffle/common/config/RssBaseConf.java:303]: I have changed the code as you suggested.

**Review by jerqi — APPROVED** [review]
LGTM.

**Comment by jerqi** [comment]
@lwllvyb Could you add the document for this pull request?

**Comment by lwllvyb** [comment]
> @lwllvyb Could you add the document for this pull request?

I will create a new PR to add the document.

**Pull request merged**